src/unity/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -145,19 +146,49 @@ void activity_classifier::init_options(
   add_or_update_state(flexmap_to_varmap(options.current_option_values()));
 }
 
+std::tuple<float, float> activity_classifier::compute_validation_metrics(
+    data_iterator *validation_data_iterator_, size_t prediction_window,
+    size_t num_classes, size_t batch_size) {
+  float cumulative_val_loss = 0.f;
+  size_t val_size = 0;
+  float cumulative_val_accuracy = 0.f;
+  validation_data_iterator_->reset();
+
+  while (validation_data_iterator_->has_next_batch()) {
+    data_iterator::batch val =
+        validation_data_iterator_->next_batch(batch_size);
+    std::map<std::string, shared_float_array> results =
+        training_model_->predict({
+            {"input", val.features},
+            {"labels", val.labels},
+            {"weights", val.weights},
+        });
+
+    const shared_float_array &output = results.at("output");
+    shared_float_array loss = results.at("loss");
```

Comment:
I feel like `output` and `loss` should either both be values (`shared_float_array`) or references (`const shared_float_array&`)